src/render.rs

```diff
@@ -296,6 +226,206 @@ struct DebugQuad {
     size: [i32; 2],
 }
 
+/// Source code for a single GLSL shader.
+pub(crate) struct ShaderSource {
+    /// The shader source code.
+    code: String,
+}
+
+/// Source code for a single GLSL program.
+pub(crate) struct ProgramSource {
+    /// Vertex shader source code.
+    pub vs: ShaderSource,
+
+    /// Fragment/pixel shader source code.
+    pub ps: ShaderSource,
+}
+
+/// Source code for every GLSL program used by `three`.
+pub(crate) struct ProgramSourceSet {
+    /// Source code for the basic pipeline.
+    pub basic: ProgramSource,
+
+    /// Source code for the Gouraud pipeline.
+    pub gouraud: ProgramSource,
+
+    /// Source code for the PBR pipeline.
+    pub pbr: ProgramSource,
+
+    /// Source code for the Phong pipeline.
+    pub phong: ProgramSource,
+
+    /// Source code for the quad pipeline.
+    pub quad: ProgramSource,
+
+    /// Source code for the shadow casting pipeline.
+    pub shadow: ProgramSource,
+
+    /// Source code for the skybox background pipeline.
+    pub skybox: ProgramSource,
+
+    /// Source code for the sprite pipeline.
+    pub sprite: ProgramSource,
+}    
+
+/// All pipeline state objects used by the `three` renderer.
+struct PipelineStates {
+    /// Corresponds to `Material::LineBasic`.
+    line_basic: BasicPipelineState,
+
+    /// Corresponds to `Material::MeshBasic` with `wireframe` set to `false`.
+    mesh_basic_fill: BasicPipelineState,
+
+    /// Corresponds to `Material::MeshBasic` with `wireframe` set to `true`.
+    mesh_basic_wireframe: BasicPipelineState,
+
+    /// Corresponds to `Material::MeshGouraud`.
+    mesh_gouraud: BasicPipelineState,
+
+    /// Corresponds to `Material::MeshPhong`.
+    mesh_phong: BasicPipelineState,
+
+    /// Corresponds to `Material::Sprite`.
+    sprite: BasicPipelineState,
+
+    /// Used internally for shadow casting.
+    shadow: gfx::PipelineState<back::Resources, shadow_pipe::Meta>,
+
+    /// Used internally for rendering sprites.
+    quad: gfx::PipelineState<back::Resources, quad_pipe::Meta>,
+
+    /// Corresponds to `Material::MeshPbr`.
+    pbr: gfx::PipelineState<back::Resources, pbr_pipe::Meta>,
+
+    /// Used internally for rendering `Background::Skybox`.
+    skybox: gfx::PipelineState<back::Resources, quad_pipe::Meta>,
+}
+
+impl PipelineStates {
+    pub fn init(src: &ProgramSourceSet, factory: &mut back::Factory) -> Self {
```

Comment:
The idea here is if the `ProgramSourceSet` needs to change then we can reload *all* the shaders in one go.